Expose all endpoints for a given instance

oyving opened the pull request:
FYI @ldalves

Commit:
Expose all endpoints for a given instance

mpolden commented on controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/routing/RoutingApiHandler.java:
```diff
@@ -85,11 +86,38 @@ private HttpResponse get(Path path, HttpRequest request) {
+                .distinct()
+                .collect(Collectors.toMap(
+                        zone -> zone,
+                        zone -> controller.routing().globalRotationStatus(new DeploymentId(instanceId, zone))
```
This will only get status for endpoints having rotations. Have a look at this method: https://github.com/vespa-engine/vespa/blob/a8c616b18803e2e3a959b9ba987deb64fabf252a/controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/routing/RoutingApiHandler.java#L258-L306

Try extracting a method from this and reuse it here.

Commit:
Try to generalise from the toSlime() method

There are multiple types of endpoints and routing methods. Here we
try to use the same logic as the toSlime() method to generate the
status JSON from the instance/endpoint POV.

Not sure if the JSON format makes sense or if are in a position where
we can get multiple contradictory statuses per endpoint+zone.

oyving commented:
Alright, take two. Trying to use same methods as the old `toSlime()` method to calculate the status. Not sure if there are some underlying constraints here that I'm ignoring and that could create confusing and contradictory output. E.g. for a given endpoint+zone, can there be two statuses that contradict each other?

mpolden commented on controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/routing/RoutingApiHandler.java:
```diff
@@ -369,4 +426,7 @@ private static String asString(RoutingMethod method) {
         }
     }
 
+    private static class RoutingStatus {
```
Unused?

oyving replied:
Indeed. Amended and removed

mpolden reviewed (commented):
> E.g. for a given endpoint+zone, can there be two statuses that contradict each other?

No, because a deployment/zone either supports shared routing (always rotation) or exclusive routing (always status from routing policy).

However, since we supported both methods at some point this constraint isn't enforced in code.

mpolden commented on controller-server/src/test/java/com/yahoo/vespa/hosted/controller/restapi/routing/responses/endpoint/endpoints.json:
```diff
@@ -0,0 +1,31 @@
+  "endpoints": [
+    {
+      "endpoint": "default",
+      "dns": "a1--t1.global.vespa.oath.cloud",
```
How is this going to be used? Perhaps better to include the URL? Otherwise console won't know about scheme or port.

oyving replied:
We aren't really going to use those values much from the console. We only care about the name and the status for the zone for now. I just added the other values for completeness.

mpolden commented on controller-server/src/test/java/com/yahoo/vespa/hosted/controller/restapi/routing/responses/endpoint/endpoints.json:
```diff
@@ -0,0 +1,31 @@
+{
+  "endpoints": [
+    {
+      "endpoint": "default",
```
Consider calling this `name` or `endpointName`.

oyving replied:
`name`, agreed.

Commit:
endpoint -> name, dns -> dnsName

ldalves commented:
Thanks!
